reloc.ml

```diff
@@ -94,21 +94,90 @@ let get_output1 ?use_bash cmd =
 
 let mk_dirs_opt pr = String.concat " " (List.map (fun s -> pr ^ (Filename.quote s)) !dirs)
 
+exception Not_utf8
+
+let utf8_next s i =
+  let fail () = raise Not_utf8 in
+  let check i =
+    if i >= String.length s then fail ();
+    let n = Char.code s.[i] in
+    if n lsr 6 <> 0b10 then fail () else n
+  in
+  if !i >= String.length s then fail ();
+  match s.[!i] with
+  | '\000'..'\127' as c ->
+      let n = Char.code c in
+      i := !i + 1;
+      n
+  | '\192'..'\223' as c ->
+      let n1 = Char.code c in
+      let n2 = check (!i+1) in
+      let n =
+        ((n1 land 0b11111) lsl 6) lor
+        ((n2 land 0b111111))
+      in
+      i := !i + 2;
+      n
+  | '\224'..'\239' as c ->
+      let n1 = Char.code c in
+      let n2 = check (!i+1) in
+      let n3 = check (!i+2) in
+      let n =
+        ((n1 land 0b1111) lsl 12) lor
+        ((n2 land 0b111111) lsl 6) lor
+        ((n3 land 0b111111))
+      in
+      i := !i + 3;
+      n
+  | '\240'..'\247' as c ->
+      let n1 = Char.code c in
+      let n2 = check (!i+1) in
+      let n3 = check (!i+2) in
+      let n4 = check (!i+3) in
+      let n =
+        ((n1 land 0b111) lsl 18) lor
+        ((n2 land 0b111111) lsl 12) lor
+        ((n3 land 0b111111) lsl 6) lor
+        ((n4 land 0b111111))
+      in
+      i := !i + 4;
+      n
+  | _ ->
+      fail ()
+
+let toutf16 s =
+  let i = ref 0 in
+  let b = Buffer.create (String.length s) in
+  let cp n = Buffer.add_char b (Char.chr (n land 0xFF)); Buffer.add_char b (Char.chr ((n lsr 8) land 0xFF)) in
+  while !i < String.length s do
+    let n = utf8_next s i in
+    if n <= 0xFFFF then cp n else (cp (0xD7C0 + (n lsl 10)); cp (0xDC00 + (n land 0x3FF)))
```

Comment:
Is this supposed to be UTF-16 surrogate encoding? The formula for the high surrogate is (very) wrong, I'm afraid - although sufficiently so that I'm wondering if I'm missing something else.

```ocaml
else let n = n - 0x10000 in (cp (0xD8000 + (n lsr 10)); ...)
```

is allowing me to run `make world` from a directory called :camel:

Reply:
Ooops, did I make a mistake? I will take a closer look in a bit, thanks!

Reply:
Indeed, the first term seems to be really wrong - I can't understand what I must have been thinking to be honest... And it did work in the few tests I did -- pure luck I guess!

Reply:
Can you please make a PR with this fix? Thanks!

Reply:
Doh! Just the for proverbial record, it's described [on Wikipedia](https://en.wikipedia.org/wiki/UTF-16#U.2B10000_to_U.2B10FFFF).

Over my :coffee:, I couldn't resist the puzzle: your code works for the high surrogates for exactly 16 of the non-BMP characters (precisely ones where you get the extra 1 required to correct the base bit pattern for 0xD800)... perhaps very appropriately, lots of them are from Linear-B :smile: &#65537; &#65601; &#65665; &#65729; &#65793; &#65857; &#65921; &#66177; &#66241; &#66305; &#66369; &#66433; &#66497;

Reply:
See #47